pennylane/templates/state_preparations/mottonen.py

```diff
@@ -134,69 +140,79 @@ def _uniform_rotation_dagger(gate, alpha, control_wires, target_wire):
+        [(2 * j - 1) * 2 ** (k - 1) + l - 1 for l in range(1, 2 ** (k - 1) + 1)]
+        for j in range(1, 2 ** (n - k) + 1)
+    ]
+    indices2 = [
+        [(2 * j - 2) * 2 ** (k - 1) + l - 1 for l in range(1, 2 ** (k - 1) + 1)]
+        for j in range(1, 2 ** (n - k) + 1)
+    ]
 
-    for i in range(len(omega)):
-        j = int(np.ceil((i + 1) * 2 ** (-k)))
-        s_condition = 2 ** (k - 1) * (2 * j - 1)
-        s_i = 1.0 if (i + 1) > s_condition else -1.0
-        alpha_z_k[j - 1] = alpha_z_k[j - 1] + s_i * omega[i] / 2 ** (k - 1)
+    term1 = np.take(omega, indices=indices1)
+    term2 = np.take(omega, indices=indices2)
+    diff = (term1 - term2) / 2 ** (k - 1)
 
-    return alpha_z_k
+    return np.sum(diff, axis=1)
 
 
 def _get_alpha_y(a, n, k):
-    r"""Computes the rotation angles for the Y rotations applied to the k'th qubit.
+    r"""Computes the rotation angles required to implement the uniformly controlled Y rotation
+    applied to the :math:`k`th qubit.
 
-    The angles are related to a as:
+    The :math:`j`th angle is related to the absolute values a of the desired amplitudes via:
 
     .. math:: \alpha^{y,k}_j = 2 \arcsin \sqrt{ \frac{ \sum_{l=1}^{2^{k-1}} a_{(2j-1)2^{k-1} +l}^2  }{ \sum_{l=1}^{2^{k}} a_{(j-1)2^{k} +l}^2  } }
 
     Args:
-        a (float): absolute values of the input
-        n (int): total number of qubits
+        a (array): absolute values of the state to prepare
+        n (int): total number of qubits for the uniformly-controlled rotation
         k (int): index of current qubit
 
     Returns:
         array representing :math:`\alpha^{y,k}`
     """
+    indices_numerator = [
+        [(2 * (j + 1) - 1) * 2 ** (k - 1) + l for l in range(2 ** (k - 1))]
+        for j in range(2 ** (n - k))
+    ]
+    numerator = np.take(a, indices=indices_numerator)
+    numerator = np.sum(np.abs(numerator) ** 2, axis=1)
 
-    numerator = np.zeros((2 ** (n - k),), dtype=np.float64)
-    denominator = np.zeros((2 ** (n - k),), dtype=np.float64)
-    alpha = np.zeros((2 ** (n - k),), dtype=np.float64)
-
-    # compute all numerators/denominators at once for efficiency
-    for i in range(len(a)):
-        j_ = int(math.ceil((i + 1) / 2 ** k))
-        l = (i + 1) - (2 * j_ - 1) * 2 ** (k - 1)
-        is_part_numerator = 1 <= l <= 2 ** (k - 1)
-
-        if is_part_numerator:
-            numerator[j_ - 1] += a[i] * a[i]
-        denominator[j_ - 1] += a[i] * a[i]
+    indices_denominator = [
+        [j * 2 ** k + l for l in range(0, int(2 ** k))] for j in range(2 ** (n - k))
+    ]
+    denominator = np.take(a, indices=indices_denominator)
+    denominator = np.sum(np.abs(denominator) ** 2, axis=1)
 
-    for i in range(len(alpha)):
-        if denominator[i] != 0.0:
-            alpha[i] = numerator[i] / denominator[i]
+    # Divide only where denominator is zero, else leave initial value of zero.
+    # The equation guarantees that the numerator is also zero in the corresponding entries.
```

Comment:
so we expect that the correct entry should be 0 when the denominator (and numerator) is zero?

Reply:
Yes!